src/content/post/2024-04-05-rdf-shacl-and-kuzu.md

```diff
@@ -0,0 +1,343 @@
+
+```turtle
+@prefix kz: <http://kuzu.io/rdf-ex#> .
+@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
+@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
+
+kz:Waterloo a kz:City ;
+        kz:name "Waterloo"@en ;
+        kz:population 150000 .
+
+kz:Adam a kz:student ;
+    kz:livesIn kz:Waterloo ;
+    kz:name "Adam" ;
+    kz:age  30 .
+
+kz:student rdfs:subClassOf kz:person .
+
+kz:Karissa a kz:student ;
+       kz:bornIn kz:Waterloo ;
+       kz:name "Karissa" .
+
+kz:Zhang a kz:faculty ;
+     kz:name "Zhang" .
+
+kz:faculty rdfs:subClassOf kz:person .
+```
+
+The data represents an RDF graph with schema information about two students, Adam and Karissa,
+a faculty member named Zhang, and a city named Waterloo. It also specifies how these resources are related
+to one another. Note that `a` is just shorthand for `rdf:type` in the Turtle file, i.e., `kz:Adam a kz:student`
+means the same as `kz:Adam rdf:type kz:student`.
+
+Conceptually, this can be represented in RDF as follows:
+
+![](/img/rdf-shacl-kuzu/rdf-running-example.png)
+
+`kz:Adam` is an alias for the IRI `http://kuzu.io/rdf-ex#Adam`, as specified in the prefix section at the top of the Turtle file. Each resource's properties are
+represented as triples between the resource and literals. The relationships between resources are also represented as triples, such as `kz:Adam livesIn kz:Waterloo`.
+
+## Load data into a Kùzu database
+
+The following snippet shows how to ingest the RDF data into Kùzu. We first create an RDF graph
+and then copy data from the Turtle file named `uni.ttl` into a local database directory named `db`.
+
+```py
+from pathlib import Path
+import kuzu
+
+DB_DIR = "db"
+db_path = Path(DB_DIR)
+# Populate Kùzu with the RDF data
+db = kuzu.Database(DB_DIR)
+conn = kuzu.Connection(db)
+
+conn.execute("CREATE RDFGraph UniKG")
+conn.execute("COPY UniKG FROM 'uni.ttl'")
+```
+
+## Register Kùzu in RDFLib plugins
+
+[RDFLib](https://rdflib.readthedocs.io/en/stable/) is a Python library that provides a simple API for querying RDF data. It is extensible with plugins[^3], allowing it to work with
+different storage backends. In this case, we simply register the Kùzu plugin in RDFLib. For databases like Kùzu that offer on-disk persistence,
+the graph needs to first be opened before we can query it.
+
+```py
+import rdflib
+
+rdflib.plugin.register(
+    "kuzudb",
+    rdflib.store.Store,
+    "graph",
+    "PropertyGraph",
+)
+# Create an RDF graph instance
+graph = rdflib.Graph(store="kuzudb", identifier="kuzudb")
+# Open the RDF graph
+graph.open(configuration="db", create=True)
+```
+
+We can then run a simple SPARQL query via RDFLib to retrieve all triples from the RDF graph.
+
+```py
+import pandas as pd
+
+query = """
+    SELECT ?src ?rel ?dst
+    WHERE {
+        ?src ?rel ?dst .
+    }
+"""
+# Export query results to a Pandas DataFrame
+df = pd.DataFrame([r.asdict() for r in graph.query(query)])
```

Comment:
Interesting. How does RDFLib return the results of SPARQL queries over data residing in Kuzu? I am assuming that someone or you must have implemented a function somewhere in the "Kuzu plugin" that just gives all the triples to RDFLib. So wrote that? Maybe I'm just lost here but I think we should say a bit more about how this plugin works and what SPARQL queries can run over data in Kuzu using RDFLib. Because if it is the case that there is a way to run any SPARQL over data residing in Kuzu through RDFLib, we should be clear about it and in fact promote it.